src/meta_memcache/base/connection_pool.py

```diff
@@ -0,0 +1,118 @@
+import itertools
+import socket
+import time
+from contextlib import contextmanager
+from queue import Empty, Full, Queue
+from typing import Callable, Generator, NamedTuple, Optional
+
+from meta_memcache.base.memcache_socket import MemcacheSocket
+from meta_memcache.errors import MemcacheServerError
+from meta_memcache.settings import DEFAULT_MARK_DOWN_PERIOD_S, DEFAULT_READ_BUFFER_SIZE
+
+
+class PoolCounters(NamedTuple):
+    # Available connections in the pool, ready to use
+    available: int
+    # The # of connections active, currently in use, out of the pool
+    active: int
+    # Current stablished connections (available + active)
+    stablished: int
+    # Total # of connections created. If this keeps growing
+    # might meen the pool size is too small and we are
+    # constantly needing to create new connections:
+    total_created: int
+    # Total # of connection or socket errors
+    total_errors: int
+
+
+class ConnectionPool:
+    def __init__(
+        self,
+        server: str,
+        socket_factory_fn: Callable[[], socket.socket],
+        initial_pool_size: int,
+        max_pool_size: int,
+        mark_down_period_s: float = DEFAULT_MARK_DOWN_PERIOD_S,
+        read_buffer_size: int = DEFAULT_READ_BUFFER_SIZE,
+    ) -> None:
+        self._server = server
+        self._socket_factory_fn = socket_factory_fn
+        self._initial_pool_size: int = min(initial_pool_size, max_pool_size)
+        self._max_pool_size = max_pool_size
+        self._mark_down_period_s = mark_down_period_s
+        self._created_counter: itertools.count[int] = itertools.count(start=1)
+        self._created = 0
+        self._errors_counter: itertools.count[int] = itertools.count(start=1)
+        self._errors = 0
+        self._destroyed_counter: itertools.count[int] = itertools.count(start=1)
+        self._destroyed = 0
+        self._marked_down_until: Optional[float] = None
+        self._pool: Queue[MemcacheSocket] = Queue(self._max_pool_size)
+        self._read_buffer_size = read_buffer_size
+        for _ in range(self._initial_pool_size):
+            try:
+                self._pool.put_nowait(self._create_connection())
+            except MemcacheServerError:
```

Comment:
Yeah, the queue is just created in line `50` and `self._initial_pool_size` is `min(initial_pool_size, max_pool_size)`, so it is guaranteed to be `<=` queue max size

Reply:
Yeah but my question is that if an error occurs we're could not reach the `initial_pool_size` so we end up with a Queue with less capacity than the initial value, is that ok?

Reply:
Oh, got your point. Yeah, if there are conn errors during initialization, the pool will just start empty.

Reply:
But it will still work properly, mark it down and avoid contant reconnections if keeps being down, and when it recovers, new conns will be added to the pool. It might not reach `initial_pool_size` if the workload needs less connections to operate, but that is not an issue. The initial creation is to reduce latencies when the memcache is first used. If it was down, it is inevitable to have to reconnect at some point.